BUG: Storage handler assert error on GPU

Note that the issue tracker is NOT the place for general support.  For
discussions about development, questions about usage, or any general questions,
contact us on https://discuss.xorbits.io/.
Q14:
```
Traceback (most recent call last):
  File "/home/lichengjie/workspace/xorbits/benchmarks/tpch/run_queries.py", line 1141, in <module>
    main()
  File "/home/lichengjie/workspace/xorbits/benchmarks/tpch/run_queries.py", line 1128, in main
    run_queries(
  File "/home/lichengjie/workspace/xorbits/benchmarks/tpch/run_queries.py", line 1058, in run_queries
    globals()[f"q{query:02}"](*queries_to_args[query])
  File "/home/lichengjie/workspace/xorbits/benchmarks/tpch/run_queries.py", line 122, in wrapped
    q(*args, **kwargs)
  File "/home/lichengjie/workspace/xorbits/benchmarks/tpch/run_queries.py", line 692, in q14
    print(total)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/utils.py", line 37, in inn
    return f(self, *args, **kwargs)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/core/data.py", line 261, in __str__
    run(self)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/core/execution.py", line 55, in run
    mars_execute(mars_tileables, **kwargs)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/deploy/oscar/session.py", line 1709, in execute
    return session.execute(
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/deploy/oscar/session.py", line 1526, in execute
    execution_info: ExecutionInfo = fut.result(
  File "/home/lichengjie/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 446, in result
    return self.__get_result()
  File "/home/lichengjie/miniconda3/lib/python3.9/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/deploy/oscar/session.py", line 1689, in _execute
    await execution_info
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/deploy/oscar/session.py", line 102, in wait
    return await self._aio_task
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/deploy/oscar/session.py", line 848, in _run_in_background
    raise task_result.error.with_traceback(task_result.traceback)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/task/supervisor/processor.py", line 388, in run
    await self._process_stage_chunk_graph(*stage_args)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/task/supervisor/processor.py", line 265, in _process_stage_chunk_graph
    chunk_to_result = await self._executor.execute_subtask_graph(
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/task/execution/mars/executor.py", line 203, in execute_subtask_graph
    return await stage_processor.run()
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/task/execution/mars/stage.py", line 233, in run
    return await self._run()
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/task/execution/mars/stage.py", line 253, in _run
    raise self.result.error.with_traceback(self.result.traceback)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 393, in internal_run_subtask
    subtask_info.result = await self._retry_run_subtask(
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 494, in _retry_run_subtask
    return await _retry_run(subtask, subtask_info, _run_subtask_once)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 111, in _retry_run
    raise ex
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 73, in _retry_run
    return await target_async_func(*args)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 436, in _run_subtask_once
    return await asyncio.shield(aiotask)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/subtask/api.py", line 70, in run_subtask_in_slot
    return await ref.run_subtask.options(profiling_context=profiling_context).send(
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/backends/context.py", line 204, in send
    return self._process_result_message(result)
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/backends/context.py", line 79, in _process_result_message
    raise message.as_instanceof_cause()
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/backends/pool.py", line 604, in send
    result = await self._run_coro(message.message_id, coro)
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/backends/pool.py", line 360, in _run_coro
    return await coro
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/api.py", line 123, in __on_receive__
    return await super().__on_receive__(message)  # type: ignore
  File "xoscar/core.pyx", line 524, in __on_receive__
    raise ex
  File "xoscar/core.pyx", line 516, in xoscar.core._BaseActor.__on_receive__
    return await self._handle_actor_result(result)
  File "xoscar/core.pyx", line 401, in _handle_actor_result
    task_result = await coros[0]
  File "xoscar/core.pyx", line 444, in xoscar.core._BaseActor._run_actor_async_generator
    async with self._lock:
  File "xoscar/core.pyx", line 445, in xoscar.core._BaseActor._run_actor_async_generator
    with debug_async_timeout('actor_lock_timeout',
  File "xoscar/core.pyx", line 450, in xoscar.core._BaseActor._run_actor_async_generator
    res = await gen.athrow(*res)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/subtask/worker/runner.py", line 132, in run_subtask
    result = yield self._running_processor.run(subtask, self._slot_id)
  File "xoscar/core.pyx", line 455, in xoscar.core._BaseActor._run_actor_async_generator
    res = await self._handle_actor_result(res)
  File "xoscar/core.pyx", line 375, in _handle_actor_result
    result = await result
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/backends/context.py", line 204, in send
    return self._process_result_message(result)
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/backends/context.py", line 79, in _process_result_message
    raise message.as_instanceof_cause()
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/backends/pool.py", line 604, in send
    result = await self._run_coro(message.message_id, coro)
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/backends/pool.py", line 360, in _run_coro
    return await coro
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/api.py", line 123, in __on_receive__
    return await super().__on_receive__(message)  # type: ignore
  File "xoscar/core.pyx", line 524, in __on_receive__
    raise ex
  File "xoscar/core.pyx", line 516, in xoscar.core._BaseActor.__on_receive__
    return await self._handle_actor_result(result)
  File "xoscar/core.pyx", line 401, in _handle_actor_result
    task_result = await coros[0]
  File "xoscar/core.pyx", line 444, in xoscar.core._BaseActor._run_actor_async_generator
    async with self._lock:
  File "xoscar/core.pyx", line 445, in xoscar.core._BaseActor._run_actor_async_generator
    with debug_async_timeout('actor_lock_timeout',
  File "xoscar/core.pyx", line 450, in xoscar.core._BaseActor._run_actor_async_generator
    res = await gen.athrow(*res)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/subtask/worker/processor.py", line 778, in run
    result = yield self._running_aio_task
  File "xoscar/core.pyx", line 455, in xoscar.core._BaseActor._run_actor_async_generator
    res = await self._handle_actor_result(res)
  File "xoscar/core.pyx", line 375, in _handle_actor_result
    result = await result
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/subtask/worker/processor.py", line 601, in run
    ) = await self._store_data(chunk_graph)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/subtask/worker/processor.py", line 334, in _store_data
    store_infos = await put_infos
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/batch.py", line 149, in _async_batch
    return await self.batch_func(args_list, kwargs_list)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/storage/api/oscar.py", line 130, in batch_put
    return await self._storage_handler_ref.put.batch(*puts)
  File "xoscar/core.pyx", line 263, in __pyx_actor_method_wrapper
    async with lock:
  File "xoscar/core.pyx", line 266, in xoscar.core.__pyx_actor_method_wrapper
    result = await result
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/batch.py", line 149, in _async_batch
    return await self.batch_func(args_list, kwargs_list)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/storage/handler.py", line 246, in batch_put
    assert last_level == level
AssertionError: [address=127.0.0.1:38837, pid=742577]
```

Q14 shows the error on storage batch ``put``.
Here's an example for showing the error on storage batch ``fetch``.
```
def test(setup_gpu):
    rs = np.random.RandomState(0)
    df = md.DataFrame(
        {
            "c": [pd.Timestamp(rs.randint(1604000000, 1604481373)) for _ in range(10)],
        },
    ).to_gpu()
    r = (df["c"] > to_datetime("2000-01-01")) & (df["c"] < to_datetime("2021-01-01"))
    print(r.execute())
```

Error:
```
ERROR    xorbits._mars.services.scheduling.worker.execution:execution.py:133 Failed to run subtask OpZI6IOXspMGEJEzlrszSQd3 on band gpu-0
Traceback (most recent call last):
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 370, in internal_run_subtask
    await asyncio.wait_for(
  File "/home/lichengjie/miniconda3/lib/python3.9/asyncio/tasks.py", line 479, in wait_for
    return fut.result()
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 111, in _retry_run
    raise ex
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 73, in _retry_run
    return await target_async_func(*args)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 213, in _prepare_input_data
    await storage_api.fetch.batch(*queries)
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/batch.py", line 149, in _async_batch
    return await self.batch_func(args_list, kwargs_list)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/storage/api/oscar.py", line 214, in batch_fetch
    assert extracted_args == (level, band_name, dest_address, error)
AssertionError
ERROR    xorbits._mars.services.task.execution.mars.stage:stage.py:169 Subtask OpZI6IOXspMGEJEzlrszSQd3 errored
Traceback (most recent call last):
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 370, in internal_run_subtask
    await asyncio.wait_for(
  File "/home/lichengjie/miniconda3/lib/python3.9/asyncio/tasks.py", line 479, in wait_for
    return fut.result()
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 111, in _retry_run
    raise ex
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 73, in _retry_run
    return await target_async_func(*args)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 213, in _prepare_input_data
    await storage_api.fetch.batch(*queries)
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/batch.py", line 149, in _async_batch
    return await self.batch_func(args_list, kwargs_list)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/storage/api/oscar.py", line 214, in batch_fetch
    assert extracted_args == (level, band_name, dest_address, error)
AssertionError
ERROR    xorbits._mars.services.scheduling.worker.execution:execution.py:133 Failed to run subtask Zszn0cFeJDyjGmvWOwnrt8Vf on band gpu-0
Traceback (most recent call last):
  File "/home/lichengjie/miniconda3/lib/python3.9/asyncio/tasks.py", line 466, in wait_for
    await waiter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 370, in internal_run_subtask
    await asyncio.wait_for(
  File "/home/lichengjie/miniconda3/lib/python3.9/asyncio/tasks.py", line 469, in wait_for
    return fut.result()
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 111, in _retry_run
    raise ex
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 73, in _retry_run
    return await target_async_func(*args)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/scheduling/worker/execution.py", line 213, in _prepare_input_data
    await storage_api.fetch.batch(*queries)
  File "/home/lichengjie/miniconda3/lib/python3.9/site-packages/xoscar/batch.py", line 149, in _async_batch
    return await self.batch_func(args_list, kwargs_list)
  File "/home/lichengjie/workspace/xorbits/python/xorbits/_mars/services/storage/api/oscar.py", line 214, in batch_fetch
    assert extracted_args == (level, band_name, dest_address, error)
AssertionError

xorbits/_mars/dataframe/arithmetic/tests/test_arithmetic_execution.py:1011 (test_55)
setup_gpu = <function _new_test_session at 0x7fba7ff33820>

    def test_55(setup_gpu):
        rs = np.random.RandomState(0)
        df = md.DataFrame(
            {
                "c": [pd.Timestamp(rs.randint(1604000000, 1604481373)) for _ in range(10)],
            },
        ).to_gpu()
        r = (df["c"] > to_datetime("2000-01-01")) & (df["c"] < to_datetime("2021-01-01"))
>       print(r.execute())

_mars/dataframe/arithmetic/tests/test_arithmetic_execution.py:1020: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
_mars/core/entity/tileables.py:405: in execute
    result = self.data.execute(session=session, **kw)
_mars/core/entity/executable.py:152: in execute
    return execute(self, session=session, **kw)
_mars/deploy/oscar/session.py:1709: in execute
    return session.execute(
_mars/deploy/oscar/session.py:1526: in execute
    execution_info: ExecutionInfo = fut.result(
../../../../miniconda3/lib/python3.9/concurrent/futures/_base.py:446: in result
    return self.__get_result()
../../../../miniconda3/lib/python3.9/concurrent/futures/_base.py:391: in __get_result
    raise self._exception
_mars/deploy/oscar/session.py:1689: in _execute
    await execution_info
../../../../miniconda3/lib/python3.9/asyncio/tasks.py:688: in _wrap_awaitable
    return (yield from awaitable.__await__())
_mars/deploy/oscar/session.py:102: in wait
    return await self._aio_task
_mars/deploy/oscar/session.py:848: in _run_in_background
    raise task_result.error.with_traceback(task_result.traceback)
_mars/services/task/supervisor/processor.py:388: in run
    await self._process_stage_chunk_graph(*stage_args)
_mars/services/task/supervisor/processor.py:265: in _process_stage_chunk_graph
    chunk_to_result = await self._executor.execute_subtask_graph(
_mars/services/task/execution/mars/executor.py:203: in execute_subtask_graph
    return await stage_processor.run()
_mars/services/task/execution/mars/stage.py:233: in run
    return await self._run()
_mars/services/task/execution/mars/stage.py:253: in _run
    raise self.result.error.with_traceback(self.result.traceback)
_mars/services/scheduling/worker/execution.py:370: in internal_run_subtask
    await asyncio.wait_for(
../../../../miniconda3/lib/python3.9/asyncio/tasks.py:479: in wait_for
    return fut.result()
_mars/services/scheduling/worker/execution.py:111: in _retry_run
    raise ex
_mars/services/scheduling/worker/execution.py:73: in _retry_run
    return await target_async_func(*args)
_mars/services/scheduling/worker/execution.py:213: in _prepare_input_data
    await storage_api.fetch.batch(*queries)
../../../../miniconda3/lib/python3.9/site-packages/xoscar/batch.py:149: in _async_batch
    return await self.batch_func(args_list, kwargs_list)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <xorbits._mars.services.storage.api.oscar.StorageAPI object at 0x7fba1153b670>
args_list = [('c009f3b1a673128b315e756cdf1ecd4e_0',), ('fd0f861971c1890c79e1094b3076dde3_0',)]
kwargs_list = [{'band_name': 'gpu-0'}, {'band_name': 'numa-0'}]

    @fetch.batch
    async def batch_fetch(self, args_list, kwargs_list):
        extracted_args = []
        data_keys = []
        for args, kwargs in zip(args_list, kwargs_list):
            data_key, level, band_name, dest_address, error = self.fetch.bind(
                *args, **kwargs
            )
            if extracted_args:
>               assert extracted_args == (level, band_name, dest_address, error)
E               AssertionError

_mars/services/storage/api/oscar.py:214: AssertionError
```

--- a/python/xorbits/_mars/services/scheduling/worker/execution.py
+++ b/python/xorbits/_mars/services/scheduling/worker/execution.py
@@ -21,7 +21,7 @@ import pprint
 import sys
 from collections import defaultdict
 from dataclasses import dataclass, field
-from typing import Dict, Optional
+from typing import Dict, List, Optional
 
 import xoscar as mo
 from xoscar.errors import XoscarError
@@ -185,9 +185,9 @@ class SubtaskExecutionActor(mo.StatelessActor):
         return await mo.actor_ref(QuotaActor.gen_uid(band), address=self.address)
 
     async def _prepare_input_data(self, subtask: Subtask, band_name: str):
-        band_to_queries = defaultdict(list)
-        band_to_shuffle_queries = defaultdict(list)
-        band_to_storage_api = {}
+        band_to_queries: Dict[str, List] = defaultdict(list)
+        band_to_shuffle_queries: Dict[str, List] = defaultdict(list)
+        band_to_storage_api: Dict[str, StorageAPI] = {}
         chunk_key_to_data_keys = get_chunk_key_to_data_keys(subtask.chunk_graph)
         for chunk in subtask.chunk_graph:
             if chunk.key in subtask.pure_depend_keys:
@@ -215,18 +215,10 @@ class SubtaskExecutionActor(mo.StatelessActor):
                             key, band_name=to_fetch_band, error="ignore"
                         )
                     )
-        # if queries:
-        #     await storage_api.fetch.batch(*queries)
 
         for band, queries in band_to_queries.items():
             storage_api = band_to_storage_api[band]
             await storage_api.fetch.batch(*queries)
-
-        # if shuffle_queries:
-        #     # TODO(hks): The batch method doesn't accept different error arguments,
-        #     #  combine them when it can.
-        #
-        #     await storage_api.fetch.batch(*shuffle_queries)
 
         # TODO(hks): The batch method doesn't accept different error arguments,
         #  combine them when it can.

--- a/python/xorbits/_mars/services/scheduling/worker/execution.py
+++ b/python/xorbits/_mars/services/scheduling/worker/execution.py
@@ -185,11 +185,9 @@ class SubtaskExecutionActor(mo.StatelessActor):
         return await mo.actor_ref(QuotaActor.gen_uid(band), address=self.address)
 
     async def _prepare_input_data(self, subtask: Subtask, band_name: str):
-        queries = []
-        shuffle_queries = []
-        storage_api = await StorageAPI.create(
-            subtask.session_id, address=self.address, band_name=band_name
-        )
+        band_to_queries = defaultdict(list)
+        band_to_shuffle_queries = defaultdict(list)
+        band_to_storage_api = {}
         chunk_key_to_data_keys = get_chunk_key_to_data_keys(subtask.chunk_graph)
         for chunk in subtask.chunk_graph:
             if chunk.key in subtask.pure_depend_keys:
@@ -198,23 +196,42 @@ class SubtaskExecutionActor(mo.StatelessActor):
                 to_fetch_band = band_name
             else:
                 to_fetch_band = "numa-0"
+
+            if not band_to_storage_api.get(to_fetch_band, None):
+                band_to_storage_api[to_fetch_band] = await StorageAPI.create(
+                    subtask.session_id, address=self.address, band_name=to_fetch_band
+                )
+
             if isinstance(chunk.op, Fetch):
-                queries.append(
-                    storage_api.fetch.delay(chunk.key, band_name=to_fetch_band)
+                band_to_queries[to_fetch_band].append(
+                    band_to_storage_api[to_fetch_band].fetch.delay(
+                        chunk.key, band_name=to_fetch_band
+                    )
                 )
             elif isinstance(chunk.op, FetchShuffle):
                 for key in chunk_key_to_data_keys[chunk.key]:
-                    shuffle_queries.append(
-                        storage_api.fetch.delay(
+                    band_to_shuffle_queries[to_fetch_band].append(
+                        band_to_storage_api[to_fetch_band].fetch.delay(
                             key, band_name=to_fetch_band, error="ignore"
                         )
                     )
-        if queries:
-            await storage_api.fetch.batch(*queries)
-        if shuffle_queries:
-            # TODO(hks): The batch method doesn't accept different error arguments,
-            #  combine them when it can.
+        # if queries:
+        #     await storage_api.fetch.batch(*queries)
 
+        for band, queries in band_to_queries.items():
+            storage_api = band_to_storage_api[band]
+            await storage_api.fetch.batch(*queries)
+
+        # if shuffle_queries:
+        #     # TODO(hks): The batch method doesn't accept different error arguments,
+        #     #  combine them when it can.
+        #
+        #     await storage_api.fetch.batch(*shuffle_queries)
+
+        # TODO(hks): The batch method doesn't accept different error arguments,
+        #  combine them when it can.
+        for band, shuffle_queries in band_to_shuffle_queries.items():
+            storage_api = band_to_storage_api[band]
             await storage_api.fetch.batch(*shuffle_queries)
 
     async def _collect_input_sizes(

--- a/python/xorbits/_mars/services/scheduling/worker/tests/test_execution.py
+++ b/python/xorbits/_mars/services/scheduling/worker/tests/test_execution.py
@@ -14,6 +14,7 @@
 # limitations under the License.
 
 import asyncio
+import datetime
 import os
 import tempfile
 import time
@@ -26,6 +27,7 @@ import pandas as pd
 import pytest
 import xoscar as mo
 
+from ..... import dataframe as md
 from ..... import remote as mr
 from .....core import (
     ChunkGraph,
@@ -34,11 +36,13 @@ from .....core import (
     TileableGraph,
     TileableGraphBuilder,
 )
+from .....dataframe import to_datetime
 from .....oscar import create_actor_pool
 from .....remote.core import RemoteFunction
 from .....resource import Resource
 from .....tensor.arithmetic import TensorTreeAdd
 from .....tensor.fetch import TensorFetch
+from .....tests.core import require_cudf
 from .....utils import Timer
 from ....cluster import MockClusterAPI
 from ....lifecycle import MockLifecycleAPI
@@ -569,3 +573,41 @@ async def test_cancel_without_kill(actor_pool):
     if os.path.exists(executed_file):
         assert await storage_api.get(remote_result.key)
         os.unlink(executed_file)
+
+
+def _generate_params_for_gpu():
+    for data_type in ("df", "series"):
+        for chunked in (False, True):
+            yield data_type, chunked
+
+
+@require_cudf
+@pytest.mark.parametrize(
+    "data_type,chunked",
+    _generate_params_for_gpu(),
+)
+def test_fetch_data_from_both_cpu_and_gpu(data_type, chunked, setup_gpu):
+    chunk_size = 3 if chunked else None
+    pd_date = pd.Timestamp("1998-01-01")
+    # data from CPU
+    mars_date = to_datetime("1998-01-01")
+
+    data1 = [datetime.datetime(1989, 1, 1), datetime.datetime(2001, 1, 1)]
+    data2 = [datetime.datetime(1990, 1, 1), datetime.datetime(2021, 1, 1)]
+
+    if data_type == "df":
+        data = {"date1": data1, "date2": data2}
+        pandas_data = pd.DataFrame(data)
+        mars_data = md.DataFrame(data, chunk_size=chunk_size).to_gpu()
+    else:
+        data = data1 + data2
+        pandas_data = pd.Series(data)
+        mars_data = md.Series(data, chunk_size=chunk_size).to_gpu()
+
+    actual = mars_data >= mars_date
+    expected = pandas_data >= pd_date
+
+    if isinstance(expected, pd.DataFrame):
+        pd.testing.assert_frame_equal(expected, actual.execute().fetch(to_cpu=True))
+    else:
+        pd.testing.assert_series_equal(expected, actual.execute().fetch(to_cpu=True))

--- a/python/xorbits/_mars/services/subtask/worker/processor.py
+++ b/python/xorbits/_mars/services/subtask/worker/processor.py
@@ -359,7 +359,7 @@ class SubtaskProcessor:
                         self.subtask.subtask_id,
                         storage_level,
                     )
-                except asyncio.CancelledError:
+                except asyncio.CancelledError:  # pragma: no cover
                     logger.debug(
                         f"Cancelling put {data_info_fmt}",
                         stored_keys,

--- a/python/xorbits/_mars/services/subtask/worker/processor.py
+++ b/python/xorbits/_mars/services/subtask/worker/processor.py
@@ -27,8 +27,11 @@ from xoscar.serialization import AioSerializer
 from ....core import ChunkGraph, ExecutionError, OperandType, enter_mode
 from ....core.context import get_context
 from ....core.operand import Fetch, FetchShuffle, execute
+from ....dataframe.utils import is_cudf
 from ....lib.aio import alru_cache
 from ....optimization.physical import optimize
+from ....storage import StorageLevel
+from ....tensor.array_utils import is_cupy
 from ....typing import BandType, ChunkType
 from ....utils import Timer, calc_data_size, get_chunk_key_to_data_keys
 from ...context import ThreadedServiceContext
@@ -308,54 +311,71 @@ class SubtaskProcessor:
 
     async def _store_data(self, chunk_graph: ChunkGraph):
         # store data into storage
-        data_key_to_puts = {}
+        # The data stored in the storage needs to be separated based on levels
+        # (here we only have two levels: memory and GPU).
+        # Otherwise, it is hard to uniformly plan the storage region when writing data in batch.
+        level_to_data_key_to_puts: Dict[StorageLevel, Dict[str, Any]] = defaultdict(
+            dict
+        )
         shuffle_key_to_data = {}
         is_storage_seekable = await self._storage_api.is_seekable()
         for key, data, _ in iter_output_data(chunk_graph, self._processor_context):
             if isinstance(key, tuple) and is_storage_seekable:
                 shuffle_key_to_data[key] = data
             else:
+                storage_level = (
+                    StorageLevel.GPU
+                    if is_cudf(data) or is_cupy(data)
+                    else StorageLevel.MEMORY
+                )
                 put = self._storage_api.put.delay(key, data)
-                data_key_to_puts[key] = put
+                level_to_data_key_to_puts[storage_level][key] = put
 
-        stored_keys = list(data_key_to_puts.keys())
-        puts = data_key_to_puts.values()
-        logger.debug(
-            "Start putting data keys: %s, subtask id: %s",
-            stored_keys,
-            self.subtask.subtask_id,
-        )
+        stored_keys = []
         data_key_to_store_size = dict()
         data_key_to_memory_size = dict()
         data_key_to_object_id = dict()
-        if puts:
-            put_infos = asyncio.create_task(self._storage_api.put.batch(*puts))
-            try:
-                store_infos = await put_infos
-                for store_key, store_info in zip(stored_keys, store_infos):
-                    data_key_to_store_size[store_key] = store_info.store_size
-                    data_key_to_memory_size[store_key] = store_info.memory_size
-                    data_key_to_object_id[store_key] = store_info.object_id
-                logger.debug(
-                    "Finish putting data keys: %s, subtask id: %s",
-                    stored_keys,
-                    self.subtask.subtask_id,
-                )
-            except asyncio.CancelledError:
-                logger.debug(
-                    "Cancelling put data keys: %s, subtask id: %s",
-                    stored_keys,
-                    self.subtask.subtask_id,
-                )
-                put_infos.cancel()
+        data_info_fmt = "data keys: %s, subtask id: %s, storage level: %s"
+        for storage_level, data_key_to_puts in level_to_data_key_to_puts.items():
+            stored_keys.extend(list(data_key_to_puts.keys()))
+            puts = data_key_to_puts.values()
+            logger.debug(
+                f"Start putting {data_info_fmt}",
+                stored_keys,
+                self.subtask.subtask_id,
+                storage_level,
+            )
+            if puts:
+                put_infos = asyncio.create_task(self._storage_api.put.batch(*puts))
+                try:
+                    store_infos = await put_infos
+                    for store_key, store_info in zip(stored_keys, store_infos):
+                        data_key_to_store_size[store_key] = store_info.store_size
+                        data_key_to_memory_size[store_key] = store_info.memory_size
+                        data_key_to_object_id[store_key] = store_info.object_id
+                    logger.debug(
+                        f"Finish putting {data_info_fmt}",
+                        stored_keys,
+                        self.subtask.subtask_id,
+                        storage_level,
+                    )
+                except asyncio.CancelledError:
+                    logger.debug(
+                        f"Cancelling put {data_info_fmt}",
+                        stored_keys,
+                        self.subtask.subtask_id,
+                        storage_level,
+                    )
+                    put_infos.cancel()
 
-                logger.debug(
-                    "Cancelled put data keys: %s, subtask id: %s",
-                    stored_keys,
-                    self.subtask.subtask_id,
-                )
-                self.result.status = SubtaskStatus.cancelled
-                raise
+                    logger.debug(
+                        f"Cancelled put {data_info_fmt}",
+                        stored_keys,
+                        self.subtask.subtask_id,
+                        storage_level,
+                    )
+                    self.result.status = SubtaskStatus.cancelled
+                    raise
 
         if shuffle_key_to_data:
             await self._store_mapper_data(

